### PR TITLE
Deprecate DataModule lifecycle properties

### LINKED ISSUE #7301: [RFC] Remove has_{setup/teardown}_{stage} properties from DataModule
## 🚀 Feature

Carrying forward discussion from https://github.com/PyTorchLightning/pytorch-lightning/pull/7289#issuecomment-830193573

Remove these settings to give more control to the users and making trainer execution easier 
https://github.com/PyTorchLightning/pytorch-lightning/blob/490cc57809ebeba19003b4101393a8a058217c31/pytorch_lightning/core/datamodule.py#L33-L52

### Motivation

Consecutive calls like this have different behavior:
```
dm = MyDataModule()
model = MyLightningModule()
trainer.fit(model, datamodule=dm)
trainer.fit(model, datamodule=dm)
```

The second trainer.fit call will not run the prepare data, setup, or teardown hooks. 

Removing this check gives us these advantages
1. Simpler, consistent hook execution: More control for users for how they want to handle state changes
2. Less state management inside the trainer and datamodule: less framework overhead/maintenance for us long term and easier debuggability

cc @carmocca 

### Pitch

Remove this code:
- https://github.com/PyTorchLightning/pytorch-lightning/blob/490cc57809ebeba19003b4101393a8a058217c31/pytorch_lightning/core/datamodule.py#L27-L97
- https://github.com/PyTorchLightning/pytorch-lightning/blob/490cc57809ebeba19003b4101393a8a058217c31/pytorch_lightning/core/datamodule.py#L162-L173
- https://github.com/PyTorchLightning/pytorch-lightning/blob/490cc57809ebeba19003b4101393a8a058217c31/pytorch_lightning/core/datamodule.py#L230-L309
- https://github.com/PyTorchLightning/pytorch-lightning/blob/490cc57809ebeba19003b4101393a8a058217c31/pytorch_lightning/trainer/trainer.py#L1169-L1170
- https://github.com/PyTorchLightning/pytorch-lightning/blob/490cc57809ebeba19003b4101393a8a058217c31/pytorch_lightning/trainer/connectors/data_connector.py#L63
- https://github.com/PyTorchLightning/pytorch-lightning/blob/490cc57809ebeba19003b4101393a8a058217c31/pytorch_lightning/trainer/trainer.py#L1143-L1144

cc @carmocca

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #7301

This is the first change required, since this will be a breaking change in the API. 
First, we mark the properties as deprecated. Then in v1.6, we can remove them completely, along with all the tracking code outlined in #7301, and change the Trainer behavior to unconditionally call these hooks. 

This makes it so that the DataModule is in full control of the state, not the Trainer, giving more flexibility and predictability to users.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

**Review by kaushikb11 — APPROVED** [review]
LGTM!

**Review by carmocca — COMMENTED** [review]
These warnings will appear every time these functions are called (every setup, teardown, prepare_data call)

Considering these are properties out of the reach of users, they can get very annoying as they wouldn't be able to disable them or update their code.

What's your plan to carry out the complete deprecation?

**Comment by ananthsub** [comment]
> These warnings will appear every time these functions are called (every setup, teardown, prepare_data call)

Good catch. What do you think about the trainer directly accessing the private attribute on the module, or wrapping this in a warning cache? 

> What's your plan to carry out the complete deprecation?

After this PR is merged, in v1.6 we can remove all the tracking code and properties listed in the issue. Are there items in the issue that aren't covered which we should discuss further?

File: pytorch_lightning/core/datamodule.py [hunk 1]
```diff
@@ -160,7 +161,14 @@ def has_prepared_data(self) -> bool:
 
         Returns:
             bool: True if ``datamodule.prepare_data()`` has been called. False by default.
+
+        .. deprecated::v1.4
+            Will be removed in v1.6.0.
         """
+        rank_zero_deprecation(
+            'DataModule property `has_prepared_data` was deprecated in v1.4'
+            ' and will be removed in v1.6.'
```
Comment (Borda) [inline pytorch_lightning/core/datamodule.py:170]:
```suggestion
            'DataModule property `has_prepared_data` was deprecated in v1.4 and will be removed in v1.6.'
```

File: pytorch_lightning/core/datamodule.py [hunk 2]
```diff
@@ -169,7 +177,14 @@ def has_setup_fit(self) -> bool:
 
         Returns:
             bool: True ``if datamodule.setup(stage='fit')`` has been called. False by default.
+
+        .. deprecated::v1.4
+            Will be removed in v1.6.0.
         """
+        rank_zero_deprecation(
+            'DataModule property `has_setup_fit` was deprecated in v1.4'
+            ' and will be removed in v1.6.'
```
Comment (Borda) [inline pytorch_lightning/core/datamodule.py:186]:
```suggestion
            'DataModule property `has_setup_fit` was deprecated in v1.4 and will be removed in v1.6.'
```

File: pytorch_lightning/core/datamodule.py [hunk 3]
```diff
@@ -178,7 +193,14 @@ def has_setup_validate(self) -> bool:
 
         Returns:
             bool: True if ``datamodule.setup(stage='validate')`` has been called. False by default.
+
+        .. deprecated::v1.4
+            Will be removed in v1.6.0.
         """
+        rank_zero_deprecation(
+            'DataModule property `has_setup_validate` was deprecated in v1.4'
+            ' and will be removed in v1.6.'
```
Comment (Borda) [inline pytorch_lightning/core/datamodule.py:202]:
```suggestion
            'DataModule property `has_setup_validate` was deprecated in v1.4 and will be removed in v1.6.'
```

File: pytorch_lightning/core/datamodule.py [hunk 4]
```diff
@@ -187,7 +209,14 @@ def has_setup_test(self) -> bool:
 
         Returns:
             bool: True if ``datamodule.setup(stage='test')`` has been called. False by default.
+
+        .. deprecated::v1.4
+            Will be removed in v1.6.0.
         """
+        rank_zero_deprecation(
+            'DataModule property `has_setup_test` was deprecated in v1.4'
+            ' and will be removed in v1.6.'
```
Comment (Borda) [inline pytorch_lightning/core/datamodule.py:218]:
```suggestion
            'DataModule property `has_setup_test` was deprecated in v1.4 and will be removed in v1.6.'
```

File: pytorch_lightning/core/datamodule.py [hunk 5]
```diff
@@ -196,7 +225,14 @@ def has_setup_predict(self) -> bool:
 
         Returns:
             bool: True if ``datamodule.setup(stage='predict')`` has been called. False by default.
+
+        .. deprecated::v1.4
+            Will be removed in v1.6.0.
         """
+        rank_zero_deprecation(
+            'DataModule property `has_setup_predict` was deprecated in v1.4'
+            ' and will be removed in v1.6.'
```
Comment (Borda) [inline pytorch_lightning/core/datamodule.py:234]:
```suggestion
            'DataModule property `has_setup_predict` was deprecated in v1.4 and will be removed in v1.6.'
```

File: pytorch_lightning/core/datamodule.py [hunk 6]
```diff
@@ -205,7 +241,14 @@ def has_teardown_fit(self) -> bool:
 
         Returns:
             bool: True ``if datamodule.teardown(stage='fit')`` has been called. False by default.
+
+        .. deprecated::v1.4
+            Will be removed in v1.6.0.
         """
+        rank_zero_deprecation(
+            'DataModule property `has_teardown_fit` was deprecated in v1.4'
+            ' and will be removed in v1.6.'
```
Comment (Borda) [inline pytorch_lightning/core/datamodule.py:250]:
```suggestion
            'DataModule property `has_teardown_fit` was deprecated in v1.4 and will be removed in v1.6.'
```

File: pytorch_lightning/core/datamodule.py [hunk 7]
```diff
@@ -214,7 +257,14 @@ def has_teardown_validate(self) -> bool:
 
         Returns:
             bool: True if ``datamodule.teardown(stage='validate')`` has been called. False by default.
+
+        .. deprecated::v1.4
+            Will be removed in v1.6.0.
         """
+        rank_zero_deprecation(
+            'DataModule property `has_teardown_validate` was deprecated in v1.4'
+            ' and will be removed in v1.6.'
```
Comment (Borda) [inline pytorch_lightning/core/datamodule.py:266]:
```suggestion
            'DataModule property `has_teardown_validate` was deprecated in v1.4 and will be removed in v1.6.'
```

File: pytorch_lightning/core/datamodule.py [hunk 8]
```diff
@@ -223,7 +273,14 @@ def has_teardown_test(self) -> bool:
 
         Returns:
             bool: True if ``datamodule.teardown(stage='test')`` has been called. False by default.
+
+        .. deprecated::v1.4
+            Will be removed in v1.6.0.
         """
+        rank_zero_deprecation(
+            'DataModule property `has_teardown_test` was deprecated in v1.4'
+            ' and will be removed in v1.6.'
```
Comment (Borda) [inline pytorch_lightning/core/datamodule.py:282]:
```suggestion
            'DataModule property `has_teardown_test` was deprecated in v1.4 and will be removed in v1.6.'
```

File: pytorch_lightning/core/datamodule.py [hunk 9]
```diff
@@ -232,7 +289,14 @@ def has_teardown_predict(self) -> bool:
 
         Returns:
             bool: True if ``datamodule.teardown(stage='predict')`` has been called. False by default.
+
+        .. deprecated::v1.4
+            Will be removed in v1.6.0.
         """
+        rank_zero_deprecation(
+            'DataModule property `has_teardown_predict` was deprecated in v1.4'
+            ' and will be removed in v1.6.'
```
Comment (Borda) [inline pytorch_lightning/core/datamodule.py:298]:
```suggestion
            'DataModule property `has_teardown_predict` was deprecated in v1.4 and will be removed in v1.6.'
```

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/core/datamodule.py [hunk 10]
```diff
@@ -160,7 +161,13 @@ def has_prepared_data(self) -> bool:
 
         Returns:
             bool: True if ``datamodule.prepare_data()`` has been called. False by default.
+
+        .. deprecated::v1.4
```
Comment (justusschock) [inline pytorch_lightning/core/datamodule.py:165]: Not sure, but I thinks sphinx needs some spaces here...
```suggestion
        .. deprecated:: v1.4
```

File: pytorch_lightning/core/datamodule.py [hunk 11]
```diff
@@ -169,7 +176,13 @@ def has_setup_fit(self) -> bool:
 
         Returns:
             bool: True ``if datamodule.setup(stage='fit')`` has been called. False by default.
+
+        .. deprecated::v1.4
```
Comment (justusschock) [inline pytorch_lightning/core/datamodule.py:180]:
```suggestion
        .. deprecated:: v1.4
```

File: pytorch_lightning/core/datamodule.py [hunk 12]
```diff
@@ -178,7 +191,13 @@ def has_setup_validate(self) -> bool:
 
         Returns:
             bool: True if ``datamodule.setup(stage='validate')`` has been called. False by default.
+
+        .. deprecated::v1.4
```
Comment (justusschock) [inline pytorch_lightning/core/datamodule.py:195]:
```suggestion
        .. deprecated:: v1.4
```

File: pytorch_lightning/core/datamodule.py [hunk 13]
```diff
@@ -187,7 +206,13 @@ def has_setup_test(self) -> bool:
 
         Returns:
             bool: True if ``datamodule.setup(stage='test')`` has been called. False by default.
+
+        .. deprecated::v1.4
```
Comment (justusschock) [inline pytorch_lightning/core/datamodule.py:210]:
```suggestion
        .. deprecated:: v1.4
```

File: pytorch_lightning/core/datamodule.py [hunk 14]
```diff
@@ -196,7 +221,13 @@ def has_setup_predict(self) -> bool:
 
         Returns:
             bool: True if ``datamodule.setup(stage='predict')`` has been called. False by default.
+
+        .. deprecated::v1.4
```
Comment (justusschock) [inline pytorch_lightning/core/datamodule.py:225]:
```suggestion
        .. deprecated:: v1.4
```

File: pytorch_lightning/core/datamodule.py [hunk 15]
```diff
@@ -205,7 +236,13 @@ def has_teardown_fit(self) -> bool:
 
         Returns:
             bool: True ``if datamodule.teardown(stage='fit')`` has been called. False by default.
+
+        .. deprecated::v1.4
```
Comment (justusschock) [inline pytorch_lightning/core/datamodule.py:240]:
```suggestion
        .. deprecated:: v1.4
```

File: pytorch_lightning/core/datamodule.py [hunk 16]
```diff
@@ -214,7 +251,13 @@ def has_teardown_validate(self) -> bool:
 
         Returns:
             bool: True if ``datamodule.teardown(stage='validate')`` has been called. False by default.
+
+        .. deprecated::v1.4
```
Comment (justusschock) [inline pytorch_lightning/core/datamodule.py:255]:
```suggestion
        .. deprecated:: v1.4
```

File: pytorch_lightning/core/datamodule.py [hunk 17]
```diff
@@ -223,7 +266,13 @@ def has_teardown_test(self) -> bool:
 
         Returns:
             bool: True if ``datamodule.teardown(stage='test')`` has been called. False by default.
+
+        .. deprecated::v1.4
```
Comment (justusschock) [inline pytorch_lightning/core/datamodule.py:270]:
```suggestion
        .. deprecated:: v1.4
```

File: pytorch_lightning/core/datamodule.py [hunk 18]
```diff
@@ -232,7 +281,13 @@ def has_teardown_predict(self) -> bool:
 
         Returns:
             bool: True if ``datamodule.teardown(stage='predict')`` has been called. False by default.
+
+        .. deprecated::v1.4
```
Comment (justusschock) [inline pytorch_lightning/core/datamodule.py:285]:
```suggestion
        .. deprecated:: v1.4
```

**Review by justusschock — APPROVED** [review]

**Review by carmocca — CHANGES_REQUESTED** [review]
My point is, usually when something is deprecated, you can do two things as an user:

1. Stop using the deprecated feature
2. Upgrade to the alternative.

The issue here is that none of them can be done in this case, leading to frustration (nobody likes warnings!)

What if instead of deprecating these properties which are pretty much internal (I haven't seen any user example relying on them), we print a warning when one of these functions is called twice?

here: https://github.com/PyTorchLightning/pytorch-lightning/blob/037a71b156ac648e05735af12ddca73e48a49e57/pytorch_lightning/core/datamodule.py#L384-L385

(1) and (2) still won't be possible with this solution, but at least the number of users who are relying on the "do not call me again logic" should be much smaller than the number of users who would see these deprecation messages implemented in this PR (all datamodule users)

**Comment by ananthsub** [comment]
I talked to @carmocca around the warnings and deprecation:

1. I'll change this:
https://github.com/PyTorchLightning/pytorch-lightning/blob/037a71b156ac648e05735af12ddca73e48a49e57/pytorch_lightning/core/datamodule.py#L384-L385

to this:
```
if has run:
    rank_zero_warn(f"DataModule.{name} has already been called, so it will not be called again. In v1.6 this behavior will change to always call DataModule.{name}") 
else:
    fn(*args, **kwargs)
```

2. The warnings added to the properties will not appear during normal execution because the wrapper directly acceses the private attributes: https://github.com/PyTorchLightning/pytorch-lightning/blob/037a71b156ac648e05735af12ddca73e48a49e57/pytorch_lightning/core/datamodule.py#L372-L381

In this use case:
```
trainer.fit(module1, dm)
trainer.fit(module2, dm)
```
`setup` and `teardown` will not be called on the datamodule in the second call, which is the motivation of the issue/PR. To get the new behavior before 1.6, the user must do:
```
trainer.fit(module1, dm)
dm._has_setup_fit = False
dm._has_teardown_fit = False
trainer.fit(module2, dm)
```
It is not worth adding a `reset()` method now to clear these states because we will change the behavior in v1.6 anyways, meaning we'd deprecate this `reset()` method then.

3. @carmocca @awaelchli I'm still in favor of deprecating these properties. @carmocca raised a good point that users who are upgrading from pre Lightning1.4 to Lightning 1.6+ won't see these warnings, and their code will silently change behavior from their POV. However, we can address this via the Breaking Changes section in the release announcements and by making the execution flow easier to debug (which hooks were called when dumped to a file for inspection). Even now we consider these properties as mostly internal to the datamodule, but they are on the public API, which strongly suggests they ought to be removed.

**Comment by carmocca** [comment]
> I'm still in favor of deprecating these properties ...

How easy do we think it will be for users to write idempotent `{setup,teardown,prepare_data}` implementations?

For example, a basic user might have:
```
def prepare_data(self):
    torchvision.MNIST(root=self.data_root, download=True)
```

In this case, `prepare_data()` can be called any number of times because torchvision has already written the checks to know if the data is already available.

Now, what if you have some complex pre-processing code where knowing whether it has been applied or not is not as simple as a file existence check. In this case, the `DataModule.has_prepare_data` is helpful, otherwise, the user is left to write and manage this flag himself.

Are we okay with this?

File: pytorch_lightning/core/datamodule.py [hunk 19]
```diff
@@ -381,8 +433,13 @@ def wrapped_fn(*args: str, **kwargs: Optional[str]) -> Any:
                 has_run = obj._has_prepared_data
                 obj._has_prepared_data = True
 
-            if not has_run:
-                return fn(*args, **kwargs)
+            if has_run:
+                rank_zero_warn(
+                    f"DataModule.{name} has already been called, so it will not be called again. "
+                    f"In v1.6 this behavior will change to always call DataModule.{name}."
+                )
```
Comment (carmocca) [inline pytorch_lightning/core/datamodule.py:440]: You can take this test, move it to deprecated_api, and use it to test the warning:

https://github.com/PyTorchLightning/pytorch-lightning/blob/906c067b07a7caa148615880792e65bb779ee3c0/tests/core/test_datamodules.py#L529-L569

**Comment by ananthsub** [comment]
@carmocca the implementation can check for whether the side-effect has already taken place.

Another reason is what came up in the slack questions channel today. Users can use the datamodule outside of the Lightning Trainer: https://pytorch-lightning.readthedocs.io/en/latest/extensions/datamodules.html#datamodules-without-lightning

In these cases, `stage` can be a free-form argument, but there's no property that corresponds to it. That's an API gap which we can simplify if we remove the properties altogether

**Comment by ananthsub** [comment]
@carmocca do you think this PR is ready to go now?

**Review by carmocca — APPROVED** [review]
Sorry! Forgot I had requested changes

Great job!

**Pull request merged**